service/internal/clientpair/clientpair.go

```diff
@@ -0,0 +1,111 @@
+package clientpair
+
+import (
+	"context"
+	"strings"
+
+	"github.com/giantswarm/apiextensions-application/api/v1alpha1"
+	"github.com/giantswarm/helmclient/v4/pkg/helmclient"
+	"github.com/giantswarm/microerror"
+	"github.com/giantswarm/micrologger"
+
+	"github.com/giantswarm/chart-operator/v2/service/controller/chart/key"
+)
+
+const (
+	// privateNamespace defines GS-protected namespace the prvHelmClient
+	// is meant for. For App CRs created outside this namespace, the
+	// pubHelmClient should be used.
+	privateNamespace = "giantswarm"
+
+	// WC App Operators are the only cases where the elevated prvHelmClient
+	// has to be used outside the `giantswarm` namespace. It is due to these
+	// apps installing reources the `default:automation` does not have access to.
+	// However, existance of the App CR for unique App Operator outside the
+	// `giantswarm` namespace opens a door to scenarios described here:
+	// https://github.com/giantswarm/giantswarm/issues/22100.
+	// The appOperatorChart defines legal prefix for what elevated Helm Client
+	// can install outside the `giantswarm` namespace.
+	appOperatorChart = "https://giantswarm.github.io/control-plane-catalog/app-operator"
```

Comment:
When testing changes to `app-operator` using the `control-plane-test-catalog`, will this not be an issue?

Reply:
Depends, if we have any sort of automation that creates `app-operators` from a test catalog in WC namespaces - then yes, I think it may be a problem. I mean, this particular check is there because we have `cluster-apps-operator` creating something outside the `giantswarm` namespace, namely `app-operators` in WC namespaces, that should be reconciled by a privileged (private) Helm client, because these are installed in-cluster. Normally, outside the `giantswarm` we are going to use limited (public) client, but `app-operators` are exception to this rule we needed to account for. I don't know of any controller using the `test` catalog, so I haven't included it, but I'm not 100% sure. Anyway I would be surprised to find such controller. And for us, it doesn't pose a problem because if we need to test new `app-operator` from a `test` catalog, we can still create it, but from the `giantswarm` namespace, instead of WC namespace. I mean, the App CR would look the same, but we would change `metadata.namespace` to `giantswarm` instead of for example `demo0`.

Reply:
Oh ok, so it is `cluster-apps-operator` doing that. I was thinking when using `opsctl deploy` or stopping Flux and maybe manually updating and App CR to test a change it is till `chart-operator` that will pick up the Chart CR for it and deploy the change. In that case the NS is the WC name and the referenced App will be in the test catalog.

Reply:
It's still `chart-operator` who install the Helm Chart, but the App CR is created by `cluster-apps-operator`. In the scanrio you described,  we would need to include `control-plane-test-catalog` because you will be updating App CR that's outside the `giantswarm` namespace. It's like this:

`cluster-apps-operator` --- creates ---> `app-operator` App CR <--- watches for --- unique `app-operator` --- creates ---> `app-operator` Chart CR <---  watches for --- `chart-operator` --- installs ---> Helm Chart

Now, if there is no automation creating in-cluster App CR with `control-plane-test-catalog` we do not need to include it. But if you want to edit App CR that was created by the `cluster-apps-operator`, to change version and catalog, then we need to add the extra catalog to the set.

Another solution for testing I was referring to is about first scaling down the original `app-operator` deployment created for the WC, and then installing new one for testing with App CR created in the `giantswarm` namespace, targetting the WC namespace with `.spec.namespace` of course.